src/object/obj_tx.c

```diff
@@ -792,154 +841,815 @@ dc_tx_open(tse_task_t *task)
 }
 
 struct tx_commit_cb_args {
-	struct dc_tx	*tcca_tx;
-	crt_rpc_t	*tcca_req;
+	struct dc_tx		*tcca_tx;
+	crt_rpc_t		*tcca_req;
+	daos_tx_commit_t	*tcca_args;
 };
 
 static int
 dc_tx_commit_cb(tse_task_t *task, void *data)
 {
-	struct tx_commit_cb_args *tcca = (struct tx_commit_cb_args *)data;
+	struct tx_commit_cb_args *tcca = data;
 	struct dc_tx		 *tx = tcca->tcca_tx;
 	crt_rpc_t		 *req = tcca->tcca_req;
+	struct obj_cpd_out	 *oco = crt_reply_get(req);
+	tse_task_t		 *pool_task = NULL;
+	uint64_t		 *sub_epochs;
 	int			  rc = task->dt_result;
+	int			  rc1;
+	bool			  locked = true;
 
 	D_MUTEX_LOCK(&tx->tx_lock);
 
+	if (rc == 0) {
+		rc = oco->oco_ret;
+		if (rc == 0) {
+			int	*sub_rets = oco->oco_sub_rets.ca_arrays;
+
+			/* FIXME: currently, we pack one DTX per CPD RPC. */
+			rc = sub_rets[0];
+		}
+	}
+
 	if (rc == 0) {
 		tx->tx_status = TX_COMMITTED;
+		dc_tx_cleanup(tx);
+
+		sub_epochs = oco->oco_sub_epochs.ca_arrays;
+		if (tx->tx_epoch.oe_value == 0)
+			/* FIXME: currently, we pack one DTX per CPD RPC. */
+			tx->tx_epoch.oe_value = sub_epochs[0];
+
+		if (tx->tx_epoch.oe_value == 0)
+			D_WARN("Server forgot to reply epoch for TX "DF_DTI"\n",
+			       DP_DTI(&tx->tx_id));
+		else
+			tx->tx_epoch.oe_flags &= ~DTX_EPOCH_UNCERTAIN;
+
 		goto out;
 	}
 
 	tx->tx_status = TX_FAILED;
 
-	/* Required to restart the TX with newer epoch. */
-	if (rc == -DER_TX_RESTART)
+	if (rc != -DER_TX_RESTART && !obj_retry_error(rc)) {
+		tx->tx_retry = 0;
+
 		goto out;
+	}
 
-	/* FIXME:
-	 * 1. Check pool map, if pool map is refreshed, restart TX.
-	 * 2. Retry the RPC if it is retriable error.
-	 * 3. Failout for other failures.
-	 */
+	/* Need to refresh the local pool map. */
+	if (tx->tx_pm_ver < oco->oco_map_version) {
+		struct daos_cpd_sub_req		*dcsr;
+
+		dcsr = &tx->tx_req_cache[dc_tx_first_req(tx)];
+		tx->tx_pm_ver = oco->oco_map_version;
+		rc1 = obj_pool_query_task(tse_task2sched(task), dcsr->dcsr_obj,
+					  &pool_task);
+		if (rc1 != 0) {
+			D_ERROR("Failed to refresh the pool map: "
+				DF_RC", original error: "DF_RC"\n",
+				DP_RC(rc1), DP_RC(rc));
+			D_GOTO(out, rc = rc1);
+		}
+	}
+
+	/* Need to restart the TX with newer epoch. */
+	if (rc == -DER_TX_RESTART || rc == -DER_STALE) {
+		tx->tx_set_resend = 1;
+		sub_epochs = oco->oco_sub_epochs.ca_arrays;
+		/* FIXME: currently, we pack one DTX per CPD RPC. */
+		if (tx->tx_epoch.oe_value < sub_epochs[0])
+			tx->tx_epoch.oe_value = sub_epochs[0];
```

Comment:
This is unnecessary. The HLC Tracker is sufficient for guaranteeing that next chosen epoch will be higher than the current one.

Reply:
Then the server should not return epoch unless the DTX is successfully handled on servers.

Reply:
True, for CPD RPCs.